File: app/src/main/java/com/annapol04/munchkin/engine/Action.java
package com.annapol04.munchkin.engine;

import java.lang.reflect.Field;

public class Action {
    public static final Action NOTHING = new Action((match, game, event) -> { });
    public static final Action JOIN_PLAYER = new Action((match, game, event) -> {
            match.joinPlayer(event.getInteger());
    });
    public static final Action ASSIGN_PLAYER_NUMBER = new Action((match, game, event) -> {
        match.assignPlayerNumber(event.getScope().ordinal(), event.getInteger());
    });
    public static final Action NAME_PLAYER = new Action((match, game, event) -> {
        match.namePlayer(event.getScope().ordinal(), event.getString());
    });
    public static final Action FINISH_ROUND = new Action((match, game, event) -> {
        match.finishRound(event.getScope().ordinal());
    });
    public static final Action LEAVE_PLAYER = new Action((match, game, event) -> {
        throw new UnsupportedOperationException();
    });
    public static final Action DRAW_DOORCARD = new Action((match, game, event) -> {
        match.getPlayer(event.getScope()).drawDoorCard(Card.fromId(event.getInteger()));
    });
    public static final Action DRAW_TREASURECARD = new Action((match, game, event) -> {
        match.getPlayer(event.getScope()).drawTreasureCard(Card.fromId((event.getInteger())));
    });
    public static final Action PICKUP_CARD = new Action((match, game, event) -> {
        game.pickupCard(Card.fromId(event.getInteger()));
    });
    public static final Action PLAY_CARD = new Action((match, game, event) -> {
        game.playCard(Card.fromId(event.getInteger()));
    });
    public static final Action FIGHT_MONSTER = new Action((match, game, event) -> {
        game.fightMonster();
    });
    public static final Action RUN_AWAY = new Action((match, game, event) -> {
        game.runAwayFromMonster();
    });

    private static int idSource = 0;
    private static final Action[] lookup;

    static {
        Field[] fields = Action.class.getFields();

        lookup = new Action[fields.length];

        int i = 0;

        try {
            for (Field field : fields)
                if (field.getType() == Action.class) {
                    Action action = (Action) field.get(null);

                    action.setName(field.getName());
                    lookup[action.getId()] = action;
                }
        } catch (IllegalAccessException e) { /* wont happen... */ }
    }

    private final int id;
    private final TriConsumer<Match, Game, Event> modifier;
    private String name = null;

    public Action(TriConsumer<Match, Game, Event> modifier) {
        this.id = idSource++;
        this.modifier = modifier;
    }

    public void execute(Match match, Game game, Event data) {
        modifier.accept(match, game, data);
    }

    public int getId() {
        return id;
    }

    public static Action fromId(int id) {
        if (id >= lookup.length)
            throw new IllegalArgumentException("Invalid event action id " + id);

        return lookup[id];
    }

    private void setName(String name) {
        this.name = name;
    }

    @Override
    public String toString() {
        return name;
    }
}

File: app/src/main/java/com/annapol04/munchkin/engine/Action.kt
package com.annapol04.munchkin.engine

private var idSource = 0

class Action(private val modifier: (Match, Game, Event) -> Unit) {

    val id: Int
    private var name: String? = null

    init {
        this.id = idSource++
    }

    @Throws(IllegalEngineStateException::class)
    operator fun invoke(match: Match, game: Game, data: Event) {
        modifier(match, game, data)
    }

    private fun initialize(name: String): Action {
        this.name = name
        return this
    }

    override fun toString(): String {
        return name ?: ""
    }

    companion object {
        @JvmField val MESSAGE = Action({ match, game, event -> })
        @JvmField val JOIN_PLAYER = Action({ match, game, event -> match.joinPlayer(event.getInteger()) })
        @JvmField val ASSIGN_PLAYER_NUMBER = Action({ match, game, event -> match.assignPlayerNumber(event.scope.ordinal, event.getInteger()) })
        @JvmField val NAME_PLAYER = Action({ match, game, event -> match.namePlayer(event.scope.ordinal, event.getString()) })
        @JvmField val HAND_OVER_TOKEN = Action({ match, game, event -> match.handOverToken(event.scope, event.getInteger()) })
        @JvmField val LEAVE_PLAYER = Action({ match, game, event -> throw UnsupportedOperationException() })
        @JvmField val DRAW_DOORCARD = Action({ match, game, event -> match.drawDoorCard(event.scope, Card.fromId(event.getInteger())) })
        @JvmField val DRAW_TREASURECARD = Action({ match, game, event -> match.drawTreasureCard(event.scope, Card.fromId(event.getInteger())) })
        @JvmField val PICKUP_CARD = Action({ match, game, event -> match.getPlayer(event.scope).pickupCard(Card.fromId(event.getInteger())) })
        @JvmField val PLAY_CARD = Action({ match, game, event -> match.playCard(event.scope, Card.fromId(event.getInteger())) })
        @JvmField val FIGHT_MONSTER = Action({ match, game, event -> match.fightMonster(event.scope) })
        @JvmField val RUN_AWAY = Action({ match, game, event -> match.getPlayer(event.scope).runAway() })
        @JvmField val ENTER_TURN_PHASE = Action({ match, game, event -> match.enterTurnPhase(event.scope, TurnPhase.fromId(event.getInteger())) })

        private val lookup = Action::class.java.fields
                    .filter { it.type == Action::class.java }
                    .map { (it.get(null) as Action).initialize(it.name) }
                    .sortedBy { it.id }

        fun fromId(id: Int): Action {
            if (id >= lookup.size)
                throw IllegalArgumentException("Invalid event action id $id")

            return lookup[id]
        }
    }
}
